packages/react/src/hooks/actions/useDataStoreUpdate.ts

```diff
@@ -0,0 +1,71 @@
+import type {
+  ModelInit,
+  PersistentModel,
+  PersistentModelConstructor,
+} from '@aws-amplify/datastore';
+import { DataStore, Hub } from 'aws-amplify';
+
+import {
+  ACTION_DATASTORE_UPDATE_FINISHED_MESSAGE,
+  ACTION_DATASTORE_UPDATE_FINISHED,
+  ACTION_DATASTORE_UPDATE_STARTED_MESSAGE,
+  ACTION_DATASTORE_UPDATE_STARTED,
+  ACTIONS_CHANNEL,
+  DATASTORE_QUERY_BY_ID_ERROR,
+} from './constants';
+import { getErrorMessage } from '../../helpers/utils';
+
+export interface UseDataStoreUpdateActionOptions<
+  Model extends PersistentModel
+> {
+  model: PersistentModelConstructor<Model>;
+  id: string;
+  fields: UseDataStoreUpdateActionFields<Model>;
+}
+
+type UseDataStoreUpdateActionFields<Model> = ModelInit<
+  Model,
+  { readOnlyFields: 'createdAt' | 'updatedAt' }
+>;
+
+export const useDataStoreUpdateAction =
+  <Model extends PersistentModel>({
+    model,
+    id,
+    fields,
+  }: UseDataStoreUpdateActionOptions<Model>) =>
+  async () => {
+    try {
+      Hub.dispatch(ACTIONS_CHANNEL, {
+        event: ACTION_DATASTORE_UPDATE_STARTED,
+        data: { fields, id },
+        message: ACTION_DATASTORE_UPDATE_STARTED_MESSAGE,
+      });
+
+      const original = await DataStore.query(model, id);
+      // If query by id doesn't return an item,
+      // original will be undefined
+      // so we'll throw a helpful message.
+      if (!original) {
+        throw new Error(`${DATASTORE_QUERY_BY_ID_ERROR}: ${id}`);
```

Comment:
Should we throw an error for a query error? Maybe `console.error` would be more gentle? That way users don't need to add a `try...catch` block around it

Reply:
This will be caught by the surrounding [try/catch](https://github.com/aws-amplify/amplify-ui/pull/1253/files/f9236b03bcc9c6787f5edf216152745c6203bcbe#diff-fba6e7642d5c788d3b2c19b66c54d008f395dfe4f554d13fec405e7fa658ae48R38) and will be reported in a hub event (we won't expect the consumer to catch the error). The main reason I throw is because if original is underfined, then [Datastore.save](https://github.com/aws-amplify/amplify-ui/pull/1253/files/f9236b03bcc9c6787f5edf216152745c6203bcbe#diff-fba6e7642d5c788d3b2c19b66c54d008f395dfe4f554d13fec405e7fa658ae48R53) throws with a cryptic error that doesn't really identify the problem, which is that the original item wasn't found, so it can't be copied and updated.

Reply:
Oh, I miss that. My bad but yea that makes sense